agent_pool.go

```diff
@@ -0,0 +1,190 @@
+package tfe
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+)
+
+// Compile-time proof of interface implementation.
+var _ AgentPools = (*agentPools)(nil)
+
+// AgentPools describes all the agent pool related methods that the Terraform
+// Enterprise API supports.
+//
+// TFE API docs: https://www.terraform.io/docs/cloud/api/agents.html
+type AgentPools interface {
+	// List all the agent pools of the given organization.
+	List(ctx context.Context, organization string, options AgentPoolListOptions) (*AgentPoolList, error)
+
+	// Create a new agent pool with the given options.
+	Create(ctx context.Context, organization string, options AgentPoolCreateOptions) (*AgentPool, error)
+
+	// Read a agent pool by its ID.
+	Read(ctx context.Context, agentPoolID string) (*AgentPool, error)
+
+	// Update an agent pool by its ID.
+	Update(ctx context.Context, agentPool string, options AgentPoolUpdateOptions) (*AgentPool, error)
+
+	// Delete an agent pool by its ID.
+	Delete(ctx context.Context, agentPoolID string) error
+}
+
+// agentPools implements AgentPools.
+type agentPools struct {
+	client *Client
+}
+
+// AgentPoolList represents a list of agent pools.
+type AgentPoolList struct {
+	*Pagination
+	Items []*AgentPool
+}
+
+// AgentPool represents a Terraform Enterprise agent pool.
+type AgentPool struct {
+	ID   string `jsonapi:"primary,agent-pools"`
+	Name string `jsonapi:"attr,name"`
+}
+
+// AgentPoolListOptions represents the options for listing agent pools.
+type AgentPoolListOptions struct {
+	ListOptions
+}
+
+// List all the agent pools of the given organization. Note this currently is
+// limited to a single pool per organization (API enforced).
+func (s *agentPools) List(ctx context.Context, organization string, options AgentPoolListOptions) (*AgentPoolList, error) {
+	if !validStringID(&organization) {
+		return nil, errors.New("invalid value for organization")
+	}
+
+	u := fmt.Sprintf("organizations/%s/agent-pools", url.QueryEscape(organization))
+	req, err := s.client.newRequest("GET", u, &options)
+	if err != nil {
+		return nil, err
+	}
+
+	poolList := &AgentPoolList{}
+	err = s.client.do(ctx, req, poolList)
+	if err != nil {
+		return nil, err
+	}
+
+	return poolList, nil
+}
+
+// AgentPoolCreateOptions represents the options for creating an agent pool.
+type AgentPoolCreateOptions struct {
+	// For internal use only!
+	ID string `jsonapi:"primary,agent-pools"`
+
+	// A name to identify the agent pool.
+	Name *string `jsonapi:"attr,name"`
+}
+
+func (o AgentPoolCreateOptions) valid() error {
+	if !validString(o.Name) {
+		return errors.New("name is required")
+	}
+	return nil
+}
```

Comment:
an additional check to see if `name` contains any invalid characters would be good here. i think we do this for policy sets and workspaces, which have essentially the same naming convention. 
```suggestion
func (o AgentPoolCreateOptions) valid() error {
	if !validString(o.Name) {
		return errors.New("name is required")
	}
	if !validStringID(o.Name) {
		return errors.New("invalid value for name")
	}
	return nil
}
```

Reply:
That's what `validStringID` does 👍🏻 Not that the name implies that in any way 😑 

https://github.com/hashicorp/go-tfe/blob/72179c6bdb6484b1550110cb3ac444e929466edf/validations.go#L15-L19

Reply:
we're currently only using validString here though right? 
i was suggesting we add validStringID as well (mostly because keeping them both separate give slightly better error messages and we do it like this elsewhere)